core/payment/src/processor.rs

```diff
@@ -306,8 +311,117 @@ impl DriverRegistry {
 const REGISTRY_LOCK_TIMEOUT: Duration = Duration::from_secs(30);
 
+struct TimedMutex {
+    mutex: Mutex<DbExecutor>,
+    sender: Option<UnboundedSender<TimedMutexTaskMessage>>,
+    counter_task: Option<JoinHandle<()>>,
+}
+
+use tokio::sync::MutexGuard;
+
+enum TimedMutexTaskMessage {
+    Start(String),
+    Finish,
+}
+
+struct TimedMutexGuard<'a> {
+    mutex_guard: MutexGuard<'a, DbExecutor>,
+    sender: &'a Option<UnboundedSender<TimedMutexTaskMessage>>,
+}
+
+impl Drop for TimedMutexGuard<'_> {
+    fn drop(&mut self) {
+        if let Some(sender) = &self.sender {
+            if let Err(e) = sender.send(TimedMutexTaskMessage::Finish) {
+                log::error!("Cannot send fininsh to counter task {e}");
+            }
+        }
+    }
+}
+
+impl<'a> Deref for TimedMutexGuard<'a> {
+    type Target = MutexGuard<'a, DbExecutor>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.mutex_guard
+    }
+}
+
+impl TimedMutex {
+    fn new(db: DbExecutor) -> Self {
+        let (sender, mut receiver) =
+            tokio::sync::mpsc::unbounded_channel::<TimedMutexTaskMessage>();
+
+        let counter_task = tokio::spawn(async move {
+            loop {
+                // wait for start or close without timeout
+                let task_name = match receiver.recv().await {
+                    None => return,
+                    Some(TimedMutexTaskMessage::Start(x)) => x,
+                    Some(TimedMutexTaskMessage::Finish) => {
+                        panic!("Unexpected finish")
+                    }
+                };
+
+                let mut counter = 0;
+                loop {
+                    match tokio::time::timeout(Duration::from_secs(10), receiver.recv()).await {
+                        Err(_) => {
+                            log::error!("Long running task: {task_name}!");
+                            counter += 1;
+                            // five minutes
+                            if counter > 30 {
+                                exit(41);
+                            }
+                        }
+                        Ok(None) => panic!("Unexpected mpsc close."),
+                        Ok(Some(TimedMutexTaskMessage::Finish)) => break,
+                        Ok(Some(TimedMutexTaskMessage::Start(_))) => panic!("Unexpected start"),
+                    }
+                }
+            }
+        });
+
+        Self {
+            mutex: Mutex::new(db),
+            sender: Some(sender),
+            counter_task: Some(counter_task),
+        }
+    }
+
+    async fn timeout_lock(
+        &self,
+        duration: Duration,
+        name: &str,
+    ) -> Result<TimedMutexGuard<'_>, Elapsed> {
+        let result = tokio::time::timeout(duration, self.mutex.lock())
+            .await
+            .map_err(|e| {
+                log::info!("Failed to lock mutex in scenario {0}", name);
+                e
+            })?;
+
+        Ok(TimedMutexGuard {
```

Comment:
You seem to have forgotten to send `TimedMutexTaskMessage::Start(name)` somewhere in this function.